### DISCUSSION TITLE
`__class__` and `self.__class__` are not the same?

### DISCUSSION THREAD

**Discussion by PerchunPak** [microsoft/pyright#8305] [Q&A]
I had this code:

```python
from typing import Self

class Example:
    def clone(self) -> Self:
        return __class__()
```

and after updating pyright from 1.1.334 to 1.1.370, I started to get an error `Expression of type "Example" is incompatible with return type "Self@Example"`. However if I replace `__class__` with `self.__class__`, the error disappears.

So I am wondering, is there any difference between those two, or it is a bug in pyright?

**Comment by erictraut** [comment]
From the perspective of a type checker, `__class__` is a separate symbol from `self.__class__`. The latter is defined in typeshed's stubs (as part of the `object` class definition). The former is not, and it must be hard-coded in type checker logic.

Currently, pyright's logic for `__class__` assumes the same type as the enclosing class (in your code snippet, `Example`). I think it's reasonable to change this to `type[Self]`, which would make the above code type check without errors.

This change will be included in the next release.
